### PR TITLE
Use `vim.ui.select` for selecting lsp client

### PR DESCRIPTION
**Pull request by IndianBoy42**
#16529

### CODE REVIEW

**Comment by zeertzjq** [comment]
Same here. It's better to prefix the commit message with `feat(lsp): `.

File: runtime/lua/vim/lsp/buf.lua [hunk 1]
```diff
@@ -116,31 +116,32 @@ end
...
-    if user_choice == 0 then return nil end
-    return clients[user_choice]
+    vim.ui.select(clients, {
+      prompt = 'Select a language server:',
+      format_item = function(client)
+        return client.name
+      end,
+    }, function(item, idx)
+      on_choice(item)
+    end)
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/buf.lua:140]:
```suggestion
    }, on_choice)
```

Not necessary to wrap the function to drop an unused argument

File: runtime/lua/vim/lsp/buf.lua [hunk 2]
```diff
@@ -116,31 +116,32 @@ end
...
   elseif #clients < 1 then
-    return nil
+    on_choice(0)
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/buf.lua:142]:
```suggestion
    on_choice(nil)
```

From the docs:

```
                    {on_choice}  function ((item|nil, idx|nil) -> ()) Called
                                 once the user made a choice. `idx` is the
                                 1-based index of `item` within `item` . `nil`
                                 if the user aborted the dialog.
```

It shouldn't be called with `0` if there are no items or if the user aborted

File: runtime/lua/vim/lsp/buf.lua (hunk 1 above)
Comment (IndianBoy42) [reply runtime/lua/vim/lsp/buf.lua]: I think that was a leftover from when I was prototyping, I initially thought the return value was the index

File: runtime/lua/vim/lsp/buf.lua (hunk 2 above)
Comment (IndianBoy42) [reply runtime/lua/vim/lsp/buf.lua]: I think that was a leftover from when I was prototyping, I initially thought the return value was the index

File: runtime/lua/vim/lsp/buf.lua [hunk 3]
```diff
@@ -152,11 +151,14 @@ end
...
-  if client == nil then return end
+  select_client('textDocument/formatting', function(client)
+    if client == nil then
+      return
+    end
 
-  local params = util.make_formatting_params(options)
-  return client.request("textDocument/formatting", params, nil, vim.api.nvim_get_current_buf())
+    local params = util.make_formatting_params(options)
+    return client.request('textDocument/formatting', params, nil, vim.api.nvim_get_current_buf())
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/buf.lua:160]: I think both the `util.make_formatting_params(options)` and the `vim.api.nvim_get_current_buf()` might need to move outside of the `on_choice` callback to ensure it uses the document buffer and not the buffer or window used for the `vim.ui.select` input.

It depends a bit on whether the `vim.ui.select` implementation closes its buffer/window first before calling `on_choice`. Just to be on the safe side.

Same for the other methods.

Comment (IndianBoy42) [reply runtime/lua/vim/lsp/buf.lua]: I don't think where on_choice is run is specified anywhere, so I implemented the recommended change

**Comment by zeertzjq** [comment]
You can rebase your branch on Neovim `master` to prevent the lint failure. A lint failure was fixed in <https://github.com/neovim/neovim/commit/24f9463dd07abc5953e01d3a5ca52b734544b471>.

**Pull request merged**